rllib/agents/dqn/apex.py

```diff
@@ -283,6 +374,256 @@ def add_apex_metrics(result: dict) -> dict:
+                actors=self.workers.remote_workers(),
+                ray_wait_timeout_s=0.0,
+                max_remote_requests_in_flight_per_actor=4,
+                remote_fn=sample_and_store,
+                remote_kwargs=[{"replay_actors": self.replay_actors}]
+                * len(self.workers.remote_workers()),
+            )
+        return num_samples_ready_dict
+
+    def update_workers(self, _num_samples_ready: Dict[ActorHandle, int]) -> int:
+        """Update the remote workers that have samples ready.
+
+        Args:
+            _num_samples_ready: A mapping from ActorHandle (RolloutWorker) to
+                the number of samples returned by the remote worker.
+        Returns:
+            The number of remote workers whose weights were updated.
+        """
+        max_steps_weight_sync_delay = self.config["optimizer"]["max_weight_sync_delay"]
+        # Update our local copy of the weights if the learner thread has updated
+        # the learner worker's weights
+        if self.learner_thread.weights_updated:
+            self.learner_thread.weights_updated = False
+            weights = self.workers.local_worker().get_weights()
+            self.curr_learner_weights = ray.put(weights)
+        with self._timers[SYNCH_WORKER_WEIGHTS_TIMER]:
+            for (
+                remote_sampler_worker,
+                num_samples_collected,
+            ) in _num_samples_ready.items():
+                self.steps_since_update[remote_sampler_worker] += num_samples_collected
+                if (
+                    self.steps_since_update[remote_sampler_worker]
+                    >= max_steps_weight_sync_delay
+                ):
+                    remote_sampler_worker.set_weights.remote(
+                        self.curr_learner_weights,
+                        {"timestep": self._counters[STEPS_TRAINED_COUNTER]},
+                    )
+                    self.steps_since_update[remote_sampler_worker] = 0
+                self._counters["num_weight_syncs"] += 1
+
+    def sample_from_replay_buffer_place_on_learner_queue_non_blocking(
+        self, num_samples_collected: Dict[ActorHandle, int]
+    ) -> None:
+        """Get samples from the replay buffer and place them on the learner queue.
+
+        Args:
+            num_samples_collected: A mapping from ActorHandle (RolloutWorker) to
+                number of samples returned by the remote worker. This is used to
+                implement training intensity which is the concept of triggering a
+                certain amount of training based on the number of samples that have
+                been collected since the last time that training was triggered.
+
+        """
+        num_samples_collected = sum(num_samples_collected.values())
+        self.curr_num_samples_collected += num_samples_collected
+        if self.curr_num_samples_collected > self.config["train_batch_size"]:
```

Comment:
hey can you comment why you need to duplicate the asynchronous_parallel_requests and self.replay_sample_batches.append() logic in this if clause?
i am a bit confused.

Reply:
oh thats a pretty good question. 

So I trigger asynchronous operations to sample from the replay buffer when the number of samples collected since the last time we get samples from the replay buffer and placed them on the learner queue. But I need to be able to retrieve the results from those calls, so I have another call to `asynchronous_parallel_requests` that doesn't call any function on the replay actors, but instead checks to see if there are results from the replay actors if any remote requests were placed on them inside of that if statement.

Reply:
I'm going to add a follow up pr or do the change in my IMPALA PR for what essentially this large ray.wait cal.